### DISCUSSION TITLE
Special handling for pre-1.0.0 semver updates

### DISCUSSION THREAD

**Discussion by rarkins** [renovatebot/renovate#21656] [Suggest an Idea]
Currently we don't do any special handling for pre-1.0.0 dependencies apart from whatever is built into semver when we handle range upgrades. But importantly, pre-1.0.0 upgrades may include breaking changes, e.g. 0.2.0 -> 0.3.0 may be have multiple backwards-incompatible changes.

The main reason for considering different handling is because users may configure automerge rules based on major/minor/patch semantics, and although they may feel comfortable automerging passing minor updates post-1.0.0, they may not want to do the same pre-1.0.0.

Full requirements and solution still to be decided. Feedback welcome!

**Comment by hutson** [comment]
Using _major version zero_ versioning is a common pattern where I work, and it led me to develop [`shifted-semver-increment`](https://www.npmjs.com/package/shifted-semver-increment) to allow for incrementing versions when the major version is zero.

This aligns well with the [caret](https://github.com/npm/node-semver#caret-ranges-123-025-004) notation supported by `node-semver`, though obviously the package does not align with the behavior of `semver.inc`.

**Comment by rarkins** [comment]
@hbetts I think a lot of authors/maintainers of major version zero packages still use "usual" semver practices but maybe like having the option of "I can make non-backwards-compatible versions whenever I want without having to make it obvious via version numbers". So for Renovate to assume every minor increment of a major version zero package is breaking might lead to a worse user experience..

**Comment by rarkins** [comment]
I think this problem could be improved greatly if "the community" made it taboo to use major version zero dependencies in production. A few big companies or open source projects could shift it greatly if they approached authors of packages they depend on and said "Hi, we really like using your package but for our next major release onwards we will have a policy against major version zero dependencies. We will fork and maintain a 1.x stream of your package unless you do" then it would shift things a lot.

**Comment by hkdobrev** [comment]
> So for Renovate to assume every minor increment of a major version zero package is breaking might lead to a worse user experience..

Please, please, do keep it exactly how semver defines it - i.e. breaking changes to the API could occur between minor versions when the major version is 0. Libraries already use that when the major version is 0 and people rely on that. If renovate does something differently it would make it just more obscure and harder for people to use it who rely on a library using semver like that.

**Comment by rarkins** [comment]
@hkdobrev we are not really changing today or proposing to change anything semver defines. It's more that we allow people to set rules like "automerge" based on major/minor/patch concepts but the reality is those 3 terms mean something a little different in terms of risk when it's 0.x. They're still called major/minor/match increments in 0.x so we are not breaking the semver meaning, it's just we don't give users a way to specify different behaviour between 0.x and 1+.

One possibility we could do is keep support for major/minor/patch terminology for configuration but also add breaking/nonBreaking too. That way you could configure your automerge (to pick one example) as false for breaking and true for non-breaking. And we'd consider every upgrade for 0.x to be "breaking" hence no automerge ever. Also, this approach would allow people to choose *not* to do that too and consider patches in 0.x to be non breaking, if they really want.

Or if we want to be pedantic, we call them "maybeBreaking" and "hopefullyNotBreaking" :)

**Comment by DanielHabenicht** [comment]
Adding my support for this. 

I think we should stick to the npm semver (https://docs.npmjs.com/misc/semver#caret-ranges-123-025-004) and make it default to not automerge breaking `0.*.*` changes. 

And than add a setting like `shiftNpmBreakingChanges`. 
I don't know how this also applies to other package managers.

**Comment by fclairamb** [comment]
Dependabot handles it like what you're proposing @DanielHabenicht. Except they don't offer a way to handle it differently, which is causing the reversed issue.

The same rule applies to other packages managers. Semver has become the standard.

We could:
- Add a `"minorPost1Version"` value in the `updateTypes`, which makes it backward-compatible.
- Change the behavior of `"minor"` to make it semver-compliant (which is to behave like a major when the major is below 1) and add a `"minorAlways"`, but that means breaking backward compatibility.

By the way in my top list of problematic packages I have:
- [three.js](https://github.com/mrdoob/three.js/releases) which doesn't care about semver
- [sentry's python integration](https://github.com/getsentry/sentry-python/issues/696)

**Comment by rarkins** [comment]
I would expect that a packageRule like this can be added:

```json
{
  "packageRules": [{
    "managers": ["npm"],
    "matchCurrentVersion": "<1.0.0",
    "automerge": false
  }]
}
```

If it's added after any existing automerge rules for patch/minor then it will take precedence.

**Comment by HonkingGoose** [comment]
## My proposed default Renovate behavior

I think Renovate should never auto merge on `0.y.z` npm packages. Those packages should be merged after a review of the change log, and be merged only after a manual approval step.

I think manual labor to verify a unstable package is a feature not a bug. A user **should** do the work to verify the update.

If a user still wants to auto-merge those unstable packages by opting-in, they should be warned that this is a bad idea, and that it will probably give them a headache down the line when their project breaks due to a auto-merged update...

[Semantic Versioning 2.0.0 spec on Major version zero](https://semver.org/#spec-item-4)
> Major version zero (0.y.z) is for initial development. Anything MAY change at any time. The public API SHOULD NOT be considered stable.

## On packages staying on `0.y.z` versions

> I think this problem could be improved greatly if "the community" made it taboo to use major version zero dependencies in production. A few big companies or open source projects could shift it greatly if they approached authors of packages they depend on and said "Hi, we really like using your package but for our next major release onwards we will have a policy against major version zero dependencies. We will fork and maintain a 1.x stream of your package unless you do" then it would shift things a lot.

I really like this idea!

[Semantic Versioning 2.0.0 FAQ on releasing a 1.0.0 version](https://semver.org/#how-do-i-know-when-to-release-100)
> If your software is being used in production, it should probably already be 1.0.0. If you have a stable API on which users have come to depend, you should be 1.0.0. If you’re worrying a lot about backwards compatibility, you should probably already be 1.0.0.

**Comment by rarkins** [comment]
Workaround: add `"matchCurrentVersion": ">=1.0.0"` to any automerge rules to avoid automerging pre-1.0.0 updates

**Comment by abdonrd** [comment]
> Workaround: add `"matchCurrentVersion": ">=1.0.0"` to any automerge rules to avoid automerging pre-1.0.0 updates

@rarkins thanks for this workaround!

But I'm afraid this won't cover the case of a version wearing `~`, right?

Example with this config:
```
{
  "extends": ["config:base"],
  "rangeStrategy": "bump",
  "packageRules": [
    {
      "matchUpdateTypes": ["minor", "patch"],
      "matchCurrentVersion": ">=1.0.0",
      "automerge": true,
      "automergeType": "branch"
    }
  ]
}
```

With a package.json with: `"typescript": "~4.5.4"`

- We want branch automerge when new `~4.5.5`.
- But we want a PR with when `~4.6.0`.

For now, we will try with this:

```
"matchCurrentVersion": "!/^[~^]?0/",
```

**Comment by dominikg** [comment]
just noticed this on https://github.com/sveltejs/vite-plugin-svelte/pull/313  where cookie 0.4 -> 0.5 was included in a nonMajor batch.

Imho the semantics for nonMajor should be "leftmost non-zero number is unchanged" or to rephrase: A major update is indicated by a change in the leftmost non-zero segment.

**Comment by rarkins** [comment]
I am in favor of renovate adopting first class support for "breaking" and "non-breaking" concepts to satisfy your requirements while strongly against implementing a non-complaint definition of "major".

I'm also in favour of the community shunning pre-1.0 packages in production! Then the problem would barely exist because package authors who seek adoption will need to release a 1.x and adhere to semver semantics about breaking changes/major releases.

**Comment by dominikg** [comment]
This describes the leftmost bit for caret-ranges: https://github.com/npm/node-semver#caret-ranges-123-025-004
Not sure how official that is but it is a reality that it is being used that way today by many and also supported in other tooling.

eg, `pnpm outdated` flags cookie 0.5.0 as a major update with `"cookie":"^0.4.0"`  in package.json

**Comment by rarkins** [comment]
Semver is MAJOR.MINOR.PATCH including for 0.x. Any other interpretation is opinionated and non compliant and I prefer to not spend more time discussing it when there is already a valid proposed solution introducing the concept "breaking" which meets requirements.

**Comment by dominikg** [comment]
Sorry, i thought this issue was about handling updates in npm, not semver semantics. Also sorry for using major/breaking interchangably above.

Totally fine with a "breaking" concept, unfortunately i wasn't able to find it mentioned in the docs and a quick issue search didn't turn up anything either. I have started a [discussion](https://github.com/renovatebot/renovate/discussions/15174) to not further disturb this. (even though i am a bit confused as the title of this issue is quite literally describing what i ran into)

**Comment by rarkins** [comment]
Breaking is the concept proposed in this issue as the solution. Not redefining what "major update" means.

**Comment by ErikBrendel** [comment]
I think adding additional update types "breaking" and "nonBreaking" to renovate could solve this problem quite nicely.

My use case is to enable dependencyDashboardApproval for potentially-breaking version updates:

```js
// current approach:
{
  // works according to semver for packages >=1.0.0
  "major": {
    "dependencyDashboardApproval": true,
  },
  // workaround needed for pre-release packages:
  "packageRules": [
    {
      "matchCurrentVersion": "<1.0.0",
      "matchPackagePatterns": ["*"],
      "matchUpdateTypes": [
        "major",
        "minor",
      ],
      "dependencyDashboardApproval": true,
    }
  ],
}

// ideal solution from my point of view:
{
  // defining the setting for call kinds of "breaking" changes:
  "breaking": {
    "dependencyDashboardApproval": true,
  },
}
```

**Comment by rarkins** [comment]
I've created a new feature request issue to track this: https://github.com/renovatebot/renovate/issues/24513
